receiver/prometheusreceiver/internal/otlp_metrics_adjuster.go

```diff
@@ -0,0 +1,435 @@
+// haven't been accessed for a long period of time.
+//
+// Timeseries-level gc:
+// Some jobs that the Prometheus receiver is collecting from may export timeseries based on metrics
+// from other jobs (e.g. cAdvisor). In order to keep the timeseriesMap from leaking memory for entries
+// of no-longer existing jobs, the timeseriesMap for each job needs to remove entries that haven't
+// been accessed for a long period of time.
+//
+// The gc strategy uses a standard mark-and-sweep approach - each time a timeseriesMap is accessed,
+// it is marked. Similarly, each time a timeseriesinfo is accessed, it is also marked.
+//
+// At the end of each JobsMap.get(), if the last time the JobsMap was gc'd exceeds the 'gcInterval',
+// the JobsMap is locked and any timeseriesMaps that are unmarked are removed from the JobsMap
+// otherwise the timeseriesMap is gc'd
+//
+// The gc for the timeseriesMap is straightforward - the map is locked and, for each timeseriesinfo
+// in the map, if it has not been marked, it is removed otherwise it is unmarked.
+//
+// Alternative Strategies
+// 1. If the job-level gc doesn't run often enough, or runs too often, a separate go routine can
+//    be spawned at JobMap creation time that gc's at periodic intervals. This approach potentially
+//    adds more contention and latency to each scrape so the current approach is used. Note that
+//    the go routine will need to be cancelled upon Shutdown().
+// 2. If the gc of each timeseriesMap during the gc of the JobsMap causes too much contention,
+//    the gc of timeseriesMaps can be moved to the end of MetricsAdjuster().AdjustMetrics(). This
+//    approach requires adding 'lastGC' Time and (potentially) a gcInterval duration to
+//    timeseriesMap so the current approach is used instead.
+
+// timeseriesinfo contains the information necessary to adjust from the initial point and to detect
+// resets.
+type timeseriesinfoPdata struct {
+	mark     bool
+	initial  *pdata.Metric
+	previous *pdata.Metric
+}
+
+// timeseriesMap maps from a timeseries instance (metric * label values) to the timeseries info for
+// the instance.
+type timeseriesMapPdata struct {
+	sync.RWMutex
+	mark   bool
+	tsiMap map[string]*timeseriesinfoPdata
+}
+
+// Get the timeseriesinfo for the timeseries associated with the metric and label values.
+func (tsm *timeseriesMapPdata) get(metric *pdata.Metric, kv pdata.AttributeMap) *timeseriesinfoPdata {
+	name := metric.Name()
+	sig := getTimeseriesSignaturePdata(name, kv)
+	if metric.DataType() == pdata.MetricDataTypeHistogram {
+		// There are 2 types of Histograms whose aggregation temporality needs distinguishing:
+		// * CumulativeHistogram
+		// * GaugeHistogram
+		aggTemporality := metric.Histogram().AggregationTemporality()
+		sig += "," + aggTemporality.String()
+	}
+	tsi, ok := tsm.tsiMap[sig]
+	if !ok {
+		tsi = &timeseriesinfoPdata{}
+		tsm.tsiMap[sig] = tsi
+	}
+	tsm.mark = true
```

Comment:
Since `gc()` uses a lock to access `tsi`, I'd expect a lock here too, or otherwise atomic write. If it's already locked at a higher level, then a comment could help

Reply:
The lock is taken in `*MetricsAdjusterPdata.AdjustMetrics()`, which is the first exported method encountered when walking up the call stack from all invocations of this method.  This is the same as the OpenCensus-based implementation in metrics_adjuster.go that has been in use by this receiver for a long time.  I expect that to be the case for most of the comments on this implementation as most of them are in repeated code.

A reworking of this implementation is probably in order, particularly given the issue noted by @dashpole regarding targets that are scraped less frequently than the fixed GC interval.  That should probably be handled in a separate issue.